File: src/tps/tp1/ex13_alteracaoAleatorioC/alteracao.c
/**
 * Bernardo Vieira Rocha - 853733
 * Ciencia da Computacao PUC-MINAS
 * TP01_Q13 - Alteracao Aleatoria Recursiva (alteracaoC)
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

int isFIM(char *s) { return strcmp(s, "FIM") == 0; }

void alteracao(char palavra[], char s1, char s2, int i, int tam) {
  if (i >= tam) {
    return;
  }
  if (palavra[i] == s1) {
    palavra[i] = s2;
  }
  return alteracao(palavra, s1, s2, i + 1, tam);
}

int main(void) {
  srand(time(NULL));

  char palavra[1000];

  while (1) {
    scanf(" %[^\n]", palavra);
    if (isFIM(palavra)) {
      break;
    } else {
      char letra = (char) 'a' + rand() % 26;
      char substituto = (char) 'a' + rand() % 26;

      int tam = strlen(palavra);
      alteracao(palavra, letra, substituto, 0, tam);
      printf("%s\n", palavra);
    }
  }

  return 0;
}

File: src/tps/tp1/ex04_alteracaoAleatoria/alteracao.java
/*
  Bernardo Vieira Rocha - 853733
  Ciencia da Computacao PUC-MINAS
  TP01_Q04 - Alteracao Aleatoria (alteracao)
*/

import java.util.Random;

@SuppressWarnings("WrongPackageStatement")
public class alteracao {
  /**
   * Método que verifica se a string é "FIM".
   *
   * @param s A string a ser verificada.
   * @return true se a string for "FIM", false caso contrário.
   */
  public static boolean isFIM(String s) {
    return (s.length() == 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' &&
            s.charAt(2) == 'M');
  }

  /**
   * Método que substitui todas as ocorrências de um caractere por outro em uma
   * string.
   *
   * @param s A string na qual as substituições serão feitas.
   * @param a O caractere a ser substituído.
   * @param b O caractere substituto.
   * @return A nova string com as substituições feitas.
   */
  public static String mudar(String s, char a, char b) {
    // Cria um array de caracteres com o mesmo tamanho da string original
    char[] resultado = new char[s.length()];

    // Itera sobre cada caractere da string original
    for (int i = 0; i < s.length(); i++) {
      // Se o caractere atual for igual a 'a', substitui por 'b'
      if (s.charAt(i) == a) {
        resultado[i] = b;
      } else {
        // Caso contrário, mantém o caractere original
        resultado[i] = s.charAt(i);
      }
    }

    // Constrói a nova string a partir do array de caracteres
    String novaString = "";
    for (int i = 0; i < resultado.length; i++) {
      novaString += resultado[i];
    }

    // Retorna a nova string construída
    return novaString;
  }

  public static void main(String[] args) {
    Random gerador = new Random();
    gerador.setSeed(4);

    String[] entradas = new String[1000];
    int numEntrada = 0;

    do {
      entradas[numEntrada] = MyIO.readLine();
    } while (!isFIM(entradas[numEntrada++]));
    numEntrada--;

    for (int i = 0; i < numEntrada; ++i) {
      char a = (char)('a' + (Math.abs(gerador.nextInt()) % 26));
      char b = (char)('a' + (Math.abs(gerador.nextInt()) % 26));
      MyIO.println(mudar(entradas[i], a, b));
    }
  }
}
